### ISSUE TITLE
[PSyData] Improved design and testing for class-prefix

### ISSUE THREAD

**Issue by hiker** [stfc/PSyclone#777]
Atm you have to specify a non-default class-prefix in the apply call of a PSyData-based transformation, e.g.:
```
    data_trans = PSyDataTrans()
...
    data_trans.apply(schedule1, {"prefix": "non-default"})
...
    data_trans.apply(schedule2, {"prefix": "non-default"})
```

I think it would be more convenient if the prefix could be specified in the constructor (i.e. the same instance of a transformation would automatically create the same prefix when applied):
```
    my_data_trans = PSyDataTrans({"prefix": "non-default"})
...
    my_data_trans.apply(schedule1)
...
    my_data_trans.apply(schedule2)
```
Additionally, atm a node based on `PSyDataNode` can specify a default class-prefix that is invalid (i.e. not defined in the psyclone.cfg file) - since the test for validity is done in the `validate()` function when applying a transformation. I will add a second test for a valid prefix during `gen_code()` to catch that problem.

@arporter (since you reviewed the corresponding pull request, but anyone else feel free to comment as well) 
1. do you agree with specifying the class-prefix in the constructor?
2. Should we additionally (still) allow to overwrite the prefix in the apply function? If not, I would suggest to raise an exception if 'prefix' is specified in apply (as a way for #613 - while this would not really be a typo, but incorrect use and as such should be flagged).

PS: The reason I suggest to add an option dictionary argument to the transformation constructor instead of a normal python function argument is that long term we might want to support more default values in the constructor (e.g. the module-name could be the same for all `apply()`'s since one transformation typically will be instanced for one module, so we would only have to specify the region-name in the apply. But this is a bigger restructuring that I won't do as part of this ticket).

**Comment by arporter** [comment]
Yes, I agree. It makes sense to only have to supply the information once and I think removing it from the `apply` method is also sensible. 

The use of a dictionary sounds very like the suggestion @sergisiso made on Teams to use `**kwargs` in sub-classes. We could change the constructor to use this instead perhaps?

**Comment by hiker** [comment]
Thanks, Andy.

Well, the dictionary suggestion is what we are using for the apply/validate functions already, so to a certain degree that would break the consistency there. And I like the fact that using `options={'bla':...}` kind of documents that these options are used in the 'bowels of the transformation' somewhere (and ideally the option dictionary could become an object, which would allow us to track which options are used, so we could flag any unused option, see #613, but that would be quite a bit of change).

But tbh, we should only use one way, so if Sergi goes ahead with using **kwargs, we should probably use the same here :( Which means in order to be consistent to also change the apply/validate functions in all of our transformations and tests :(

Afaik there was the open question of documentation. I will wait how this works out, and maybe we can have a chat at the telco tomorrow about this. And I try to find out why I don't like **kwargs, but am happy using options={...} ... which is kind of the same :) I guess it's just ... if I use an option dict, the object I provide as a user at the top level is the same object that turns up down when it is used. When using **kwargs, the parameter is transformed (from keyword parameter to a dict, back to a keyword parameter). Admittedly, that is a very stupid reason ... let me think if I can come up with a better counter-argument (except the amount of work to refactor all the transforms :) ).

**Comment by sergisiso** [comment]
I don't think we have to use the same solution as both cases are different:

- The `**kwargs` suggestion was to solve the 'inconvenience' that at the moment all sub-classes have to maintain the docstring, and default value of all the inherited constructor arguments that  doesn't use itself. As an example, if we want to change the default value of Node `children=` from `None` to `[]` we would have to go to every single constructor (even outside the core PSyIR) and change it everywhere. I think this is an isolation problem as what this argument expects should only be in one place, and is becoming worse as we extend the number of PSyIR sub-classes.

- The transformation parameters dictionary have a semantic meaning when encapsulated together (are the parameters that tune some aspects of a single transformation). And they are not consumed by different constructors in the class hierarchy. The dictionary is just stored in the constructor and looked up during the transformations for certain values. In this case prefer the "options={...}" argument.

**Comment by hiker** [comment]
Considering Sergi's statement,**kwargs seems to be off the table. After looking at the constructor of other transforms (e.g. ParallelLoopTrans with takes 'schedule') other transforms already take a simple argument as parameter, so instead of starting to introduce a dict here, I'll keep it in the same style and just add one parameter for now (and face the refactoring if we should ever need it 😁 )

**Comment by arporter** [comment]
Hi @hiker, can this one be closed now?

**Comment by hiker** [comment]
I admit I am not sure - I know that you are working on the new constructor, and I had no time to check if your patch takes/took this into account. If you say yes, then please go ahead and close it. The fact that the psydata-node could specify an invalid prefix is not ideal (though tbh, you have to work quite deliberately in order to force an error this way). I can have a look tomorrow to verify this.

**Comment by arporter** [comment]
In reviewing #1224 @sergisiso suggested adding a class variable to store the prefix. This then allowed several constructors to be removed completely. In the same way, I think the same thing could be done for the `node_class` in the `PSyDataTrans` and subclasses.

**Comment by arporter** [comment]
Given that (I think) we're only talking about passing the 'prefix' argument to the transformation constructor, I think it might be better to just have it as an optional, named argument. That way support for it can just be removed from the `options` dict passed to the `apply` method. The concern about a change to the constructor interface of the base class cascading down to multiple sub-classes can be alleviated if we remove as many of those constructors as possible (which I think my first suggestion in this comment will do).

**Comment by hiker** [comment]
The reason for adding the prefix in the first place was that it was suggested we may have several psydata applications at the same time. I am not sure exactly what you mean, but if the prefix is stored in a class variable, would that still work with several different prefixes in one program unit?

**Comment by arporter** [comment]
I think so as we would give the constructor an optional `prefix` argument. So e.g. for ProfileTrans:

    def __init__(self):
        super(ProfileTrans, self).__init__(ProfileNode)

would be removed entirely and a class variable, `_node_class = ProfileNode`, would be added. The constructor of the base, PSyDataTrans class would become:

    def __init__(self, prefix=None, region=None):

as I think `prefix` and `region` (the tuple containing two strings) are common to all of the PSyData transformations.
